jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheMetadata.java

```diff
@@ -78,31 +75,24 @@ static boolean doLayerEntriesMatchMetadataEntries(
       if (layerEntries.size() != metadataEntries.size()) {
         return false;
       }
-      return pairwiseCompareAllPass(
-          layerEntries,
-          metadataEntries,
-          (layerEntry, metadataEntry) -> {
-            // Checks extraction path not equal.
-            if (!layerEntry.getExtractionPath().equals(metadataEntry.getExtractionPath())) {
-              return false;
-            }
-
-            // Checks for any source file not equal.
-            if (layerEntry.getSourceFiles().size()
-                != metadataEntry.getSourceFilesStrings().size()) {
-              return false;
-            }
-            return pairwiseCompareAllPass(
-                layerEntry.getSourceFiles(),
-                metadataEntry.getSourceFilesStrings(),
-                (sourceFile, sourceFileString) -> sourceFile.equals(Paths.get(sourceFileString)));
-          });
-    }
 
-    private static <A, B> boolean pairwiseCompareAllPass(
-        List<A> listA, List<B> listB, BiPredicate<A, B> compare) {
-      return Streams.zip(listA.stream(), listB.stream(), compare::test)
-          .allMatch(Predicate.isEqual(true));
+      // Pairwise-compares all the layer entries with the metadata layer entries.
+      Iterator<LayerEntry> layerEntriesIterator = layerEntries.iterator();
+      Iterator<LayerMetadata.LayerMetadataEntry> metadataEntriesIterator =
+          metadataEntries.iterator();
+      while (layerEntriesIterator.hasNext() && metadataEntriesIterator.hasNext()) {
+        LayerEntry layerEntry = layerEntriesIterator.next();
+        LayerMetadata.LayerMetadataEntry metadataEntry = metadataEntriesIterator.next();
+
+        boolean areSourceFilesEqual =
+            layerEntry.getSourceFileString().equals(metadataEntry.getSourceFilesString());
+        boolean areExtractionPathsEqual =
+            layerEntry.getExtractionPathString().equals(metadataEntry.getExtractionPathString());
+        if (!areSourceFilesEqual || !areExtractionPathsEqual) {
+          return false;
+        }
+      }
```

Comment:
nit: I think below will work, but not saying I want this code to be like that. Just commenting.
```java
Streams.zip(
        layerEntries.stream().map(LayerEntry::getSourceFilesString),
        metadataEntries.stream().map(LayerMetadataEntry::getSourceFilesString),
        String::equals)
   .reduce(true, Boolean::logicalAnd);
```

Reply:
Yea, I was trying to prefer using for-loops in the computationally-intensive sections (lots of computations itself and/or called many times) since streams come with a bit of overhead. For non-computationally-intensive sections, though, streams can definitely be nice to make the code flow better.

Reply:
Too bad, since the inception of Java, Java has always been slow, and we should still care about that.